samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesStoreImpl.java

```diff
@@ -0,0 +1,200 @@
+
+  /**
+   * Creates a {@link TimeSeriesStoreImpl}
+   *
+   * @param kvStore the backing kv store to use
+   * @param appendMode should the store be used in appendMode
+   */
+  public TimeSeriesStoreImpl(KeyValueStore<TimeSeriesKey<K>, V> kvStore, boolean appendMode) {
+    this.kvStore = kvStore;
+    this.appendMode = appendMode;
+  }
+
+  /**
+   * Creates a {@link TimeSeriesStoreImpl} in append mode.
+   *
+   * @param kvStore the backing kv store to use
+   */
+  public TimeSeriesStoreImpl(KeyValueStore<TimeSeriesKey<K>, V> kvStore) {
+    this(kvStore, true);
+  }
+
+  @Override
+  public void put(K key, V val, long timestamp) {
+    // For append mode, values are differentiated by an unique sequence number. For overwrite mode, the sequence
+    // number is always zero. This ensures that only the most recent value is retained.
+    if (appendMode) {
+      seqNum.getAndIncrement();
+    }
+    TimeSeriesKey<K> timeSeriesKey = new TimeSeriesKey<>(key, timestamp, seqNum.get());
+    kvStore.put(timeSeriesKey, val);
+  }
+
+  @Override
+  public ClosableIterator<TimestampedValue<V>> get(K key, long startTimestamp, long endTimestamp) {
+    validateRange(startTimestamp, endTimestamp);
+    TimeSeriesKey<K> fromKey = new TimeSeriesKey(key, startTimestamp, 0);
+    TimeSeriesKey<K> toKey = new TimeSeriesKey(key, endTimestamp, 0);
+
+    KeyValueIterator<TimeSeriesKey<K>, V> range = kvStore.range(fromKey, toKey);
+    return new TimeSeriesStoreIterator<>(range);
+  }
+
+  @Override
+  public void remove(K key, long startTimestamp, long endTimeStamp) {
+    validateRange(startTimestamp, endTimeStamp);
+    TimeSeriesKey<K> fromKey = new TimeSeriesKey(key, startTimestamp, 0);
+    TimeSeriesKey<K> toKey = new TimeSeriesKey(key, endTimeStamp, 0);
+
+    List<TimeSeriesKey<K>> keysToDelete = new LinkedList<>();
+
+    KeyValueIterator<TimeSeriesKey<K>, V> range = kvStore.range(fromKey, toKey);
+    while (range.hasNext()) {
+      keysToDelete.add(range.next().getKey());
+    }
+
+    kvStore.deleteAll(keysToDelete);
+  }
+
+  @Override
+  public void remove(K key) {
+    remove(key, 0L, Long.MAX_VALUE);
+  }
+
+  @Override
+  public void flush() {
+    kvStore.flush();
+  }
+
+  @Override
+  public void close() {
+    kvStore.close();
+  }
+
+  private void validateRange(long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+    if (startTimestamp < 0) {
+      throw new IllegalArgumentException(String.format("Start timestamp :%d is less than zero", startTimestamp));
+    }
+
+    if (endTimestamp < 0) {
+      throw new IllegalArgumentException(String.format("End timestamp :%d is less than zero", endTimestamp));
+    }
+
+    if (endTimestamp < startTimestamp) {
+      throw new IllegalArgumentException(String.format("End timestamp :%d is less than start timestamp: %d", endTimestamp, startTimestamp));
+    }
+  }
+
+  private static class TimeSeriesStoreIterator<K, V> implements ClosableIterator<TimestampedValue<V>> {
```

Comment:
as mentioned above, is it possible to automatically close it when it reaches the end?